src/CustomFunction.ts

```diff
@@ -28,115 +31,70 @@ interface FunctionFailArguments {
-    validate(callbackId, middleware);
+  public constructor(
+    callbackId: string,
+    listeners: Middleware<SlackCustomFunctionMiddlewareArgs<Options>>[],
+    options: Options,
+  ) {
+    validate(callbackId, listeners);
 
-    this.appWebClientOptions = clientOptions;
     this.callbackId = callbackId;
-    this.middleware = middleware;
-  }
-
-  public getMiddleware(): Middleware<AnyMiddlewareArgs> {
-    return async (args): Promise<void> => {
-      if (isFunctionEvent(args) && this.matchesConstraints(args)) {
-        return this.processEvent(args);
-      }
-      return args.next();
-    };
+    this.listeners = listeners;
+    this.options = options;
   }
 
-  private matchesConstraints(args: SlackCustomFunctionMiddlewareArgs): boolean {
-    return args.payload.function.callback_id === this.callbackId;
-  }
-
-  private async processEvent(args: AllCustomFunctionMiddlewareArgs): Promise<void> {
-    const functionArgs = enrichFunctionArgs(args, this.appWebClientOptions);
-    const functionMiddleware = this.getFunctionMiddleware();
-    return processFunctionMiddleware(functionArgs, functionMiddleware);
-  }
-
-  private getFunctionMiddleware(): CustomFunctionMiddleware {
-    return this.middleware;
-  }
-
-  /**
-   * Factory for `complete()` utility
-   */
-  public static createFunctionComplete(context: Context, client: WebClient): FunctionCompleteFn {
-    const token = selectToken(context);
-    const { functionExecutionId } = context;
-
-    if (!functionExecutionId) {
-      const errorMsg = 'No function_execution_id found';
-      throw new CustomFunctionCompleteSuccessError(errorMsg);
-    }
-
-    return (params: Parameters<FunctionCompleteFn>[0] = {}) =>
-      client.functions.completeSuccess({
-        token,
-        outputs: params.outputs || {},
-        function_execution_id: functionExecutionId,
-      });
-  }
-
-  /**
-   * Factory for `fail()` utility
-   */
-  public static createFunctionFail(context: Context, client: WebClient): FunctionFailFn {
-    const token = selectToken(context);
-    const { functionExecutionId } = context;
-
-    if (!functionExecutionId) {
-      const errorMsg = 'No function_execution_id found';
-      throw new CustomFunctionCompleteFailError(errorMsg);
+  public getListeners(): Middleware<AnyMiddlewareArgs>[] {
+    if (this.options.autoAcknowledge) {
+      return [
+        onlyEvents,
+        matchEventType('function_executed'),
+        matchFunction(this.callbackId),
+        autoAcknowledge,
+        ...this.listeners,
+      ] as Middleware<AnyMiddlewareArgs>[];
     }
-
-    return (params: Parameters<FunctionFailFn>[0]) => {
-      const { error } = params ?? {};
-
-      return client.functions.completeError({
-        token,
-        error,
-        function_execution_id: functionExecutionId,
-      });
-    };
+    return [
+      onlyEvents,
+      matchEventType('function_executed'),
+      matchFunction(this.callbackId),
```

Comment:
If we name this middleware "match function", the matchEventType can be merged into it. Otherwise, this middleware should be matchCallbackId()